File: python/src/apiz/_resources/account.py
from __future__ import annotations

from .._http import AsyncHttpClient, SyncHttpClient
from .._types import BalanceResponse, CheckinResponse, PackageItem, PaymentResponse


class AccountResource:
    def __init__(self, http: SyncHttpClient) -> None:
        self._http = http

    def balance(self) -> BalanceResponse:
        return BalanceResponse.model_validate(
            self._http.request("GET", "/api/v3/balance")
        )

    def checkin(self) -> CheckinResponse:
        return CheckinResponse.model_validate(
            self._http.request("POST", "/api/v3/checkin", body={})
        )

    def packages(self) -> list[PackageItem]:
        data = self._http.request("GET", "/api/v3/packages")
        return [PackageItem.model_validate(p) for p in data or []]

    def pay(self, package_id: int) -> PaymentResponse:
        return PaymentResponse.model_validate(
            self._http.request("POST", "/api/v3/pay", body={"package_id": package_id})
        )


class AsyncAccountResource:
    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def balance(self) -> BalanceResponse:
        return BalanceResponse.model_validate(
            await self._http.request("GET", "/api/v3/balance")
        )

    async def checkin(self) -> CheckinResponse:
        return CheckinResponse.model_validate(
            await self._http.request("POST", "/api/v3/checkin", body={})
        )

    async def packages(self) -> list[PackageItem]:
        data = await self._http.request("GET", "/api/v3/packages")
        return [PackageItem.model_validate(p) for p in data or []]

    async def pay(self, package_id: int) -> PaymentResponse:
        return PaymentResponse.model_validate(
            await self._http.request("POST", "/api/v3/pay", body={"package_id": package_id})
        )

File: packages/sdk/src/resources/account.ts
import type { HttpClient } from "../http.js";
import type {
  BalanceResponse,
  CheckinResponse,
  PackageItem,
  PaymentResponse,
} from "../types.js";

export class AccountResource {
  constructor(private readonly http: HttpClient) {}

  balance(): Promise<BalanceResponse> {
    return this.http.request<BalanceResponse>({
      method: "GET",
      path: "/api/v3/balance",
    });
  }

  checkin(): Promise<CheckinResponse> {
    return this.http.request<CheckinResponse>({
      method: "POST",
      path: "/api/v3/checkin",
      body: {},
    });
  }

  packages(): Promise<PackageItem[]> {
    return this.http.request<PackageItem[]>({
      method: "GET",
      path: "/api/v3/packages",
    });
  }

  pay(packageId: number): Promise<PaymentResponse> {
    return this.http.request<PaymentResponse>({
      method: "POST",
      path: "/api/v3/pay",
      body: { package_id: packageId },
    });
  }
}
